samples/resourceops/volumeop_dag.py

```diff
@@ -0,0 +1,58 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import kfp.dsl as dsl
+
+
+@dsl.pipeline(
+    name="Volume Op DAG",
+    description="The second example of the design doc."
+)
+def volume_op_dag():
+    vop = dsl.VolumeOp(
+        name="create_pvc",
+        resource_name="my-pvc",
+        size="10Gi",
+        modes=dsl.VOLUME_MODE_RWM
+    )
+
+    step1 = dsl.ContainerOp(
+        name="step1",
+        image="library/bash:4.4.23",
+        command=["sh", "-c"],
+        arguments=["echo 1 | tee /mnt/file1"],
+        volumes={"/mnt": vop.volume}
+    )
+
+    step2 = dsl.ContainerOp(
+        name="step2",
+        image="library/bash:4.4.23",
+        command=["sh", "-c"],
+        arguments=["echo 2 | tee /mnt2/file2"],
+        volumes={"/mnt2": vop.volume}
+    )
+
+    step3 = dsl.ContainerOp(
+        name="step3",
+        image="library/bash:4.4.23",
+        command=["sh", "-c"],
+        arguments=["cat /mnt/file1 /mnt/file2"],
+        volumes={"/mnt": vop.volume.after(step1, step2)}
```

Comment:
This "vop.volume.after" statement should not be needed if we use Argo's parameter passing functionality. Is there a way we could avoid exposing this feature and focus instead on using Argo's parameter passing functionality?

Reply:
> This "vop.volume.after" statement should not be needed if we use Argo's parameter passing functionality. Is there a way we could avoid exposing this feature and focus instead on using Argo's parameter passing functionality?

We already use Argo's parameter passing functionality, for resource names in general, and volume and snapshot names in particular.
We already use the existing mechanism that the compiler has to deduce dependencies based on `PipelineParam`s: Because the names of resources created by `ResourceOp`s are output `PipelineParam`s, any `ContainerOp` that needs to mount a volume will already take a dependency on the `ResourceOp` that is supposed to create it.

This mechanism however does not suffice. The problem is how to explain to the compiler that three `ContainerOp`s have to use the same volume *sequentially*, because one step needs to access files that the previous step must have placed in the volume. In this case, there could be no `PipelineParam` being passed from step to step.
  
This is our rationale: a `dsl.PipelineVolume` *is* a K8s Volume that brings dependencies, exactly as a `PipelineParam` is a string that brings dependencies. Similarly, `vop.volume.after(step1, step2, ...)` is not a statement, it is an expression that returns a new `dsl.PipelineVolume` that has taken a dependency on `step1`, `step2`.

KFP only supports `PipelineParam`s for deducing dependencies now. But in a while, we will also have artifacts. We think the platform should afford volumes the same first-class citizen treatment.

This avoids cumbersome workarounds using dummy `PipelineParams` like the following:

```python
vop = dsl.VolumeOp(...)

step1 = dsl.ContainerOp(
    name="step_1",
    image="some_image_1",
    command=["sh", "-c"],
    arguments=["echo 1 > /mnt/file1 && "
               "some_processing_that_writes_to_the_volume"],
    file_outputs={"dummy": "/mnt/file1"},
    volumes={"/mnt": vop.volume}    
)

step2 = dsl.ContainerOp(
    name="step_2",
    image="some_image_2",
    command=["sh", "-c"],
    arguments=["echo %s && some_processing_that_reads_from_volume"
               % step1.output],
    volumes={"/mnt": vop.volume}    
)
```

In that case, these Argo parameters (KFP's `PipelineParam`s) would have to be dummy data.
That, however, slows down the pipeline's execution, because Argo's executor has to retrieve and save these parameters.

It also requires the user to write additional unneeded code, just for that parameter passing.
I.e.:
  1. The user has to create a file, let's say `/tmp/dummy`
  2. Point to it in `file_outputs`
  3. Use it in another step's `command`/`argument` (e.g. `"echo %s" % previous_step.output`. This will also be logged, but it will have no value).

All of the above led us to provide such feature. That is to drive dependencies from `PipelineVolume`s (wrapper of `V1Volume`).

Reply:
Thank you @vkoukis. I am a bit hesitant about having multiple ways to specify dependencies between pipeline steps depending on how the data is stored (volume or not). 

Would it be possible to only define dependencies between steps in on way? What about a way to enforce that step_2 should execute after step_1 that does not involve volumes?

Reply:
Actually, there are already two ways of specifying dependencies between pipeline steps. One is using the `after()` method explicitly, while the other way is to use a `PipelineParam` in any field of an  `Op`.
Similarly, I assume that when KFP has support for Artifacts, the compiler will derive dependencies based on their usage.

This is what makes `PipelineParam`s powerful. I fail to see why the platform cannot afford volumes the same treatment. By not doing so, we are essentially limiting the user experience severely: If a user works with `PipelineParam`s, they can write nice, succinct code, the proposed changes carry this functionality so it works when using volumes for data exchange. We have shown in the examples, that the same kind of code can be written with volumes being the medium for passing information from step to step.

Essentially, `PipelineVolume`s should be of the same object "family" in the DSL, objects that the compiler uses to deduce dependencies.